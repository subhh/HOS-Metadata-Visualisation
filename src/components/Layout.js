import React, { useState, useEffect } from 'react'
import { Spring } from 'react-spring/renderprops.cjs'
import { useTransition, animated } from 'react-spring'
import { Box } from 'rebass'
import { debounce } from 'lodash'

const headerHeightFallback = 130
const AnimatedBox = animated(Box)

/**
 * Main layout component, upper hero section can be expanded, content area is optional
 * and adapts to the hero.
 */
const Layout = ({
  expand,
  renderHero = x => null,
  renderContent = x => null,
}) => {
  const [width, setWidth] = useState(null)
  const [height, setHeight] = useState(null)

  useEffect(() => {
    // get header height from surrounding layout
    const nav = document.querySelector('#nav-primary')
    const header = document.querySelector('#header')
    let hh

    if (nav && header) {
      hh = nav.clientHeight + header.clientHeight + 24
    } else {
      hh = headerHeightFallback
    }

    const updateDimensions = () => {
      const width = window.innerWidth
      const height = expand ? window.innerHeight - hh : window.innerHeight * 0.4
      setWidth(width)
      setHeight(height)
    }

    updateDimensions()

    const debouncedHandleResize = debounce(updateDimensions, 200)

    window.addEventListener('resize', debouncedHandleResize)

    return () => {
      window.removeEventListener('resize', debouncedHandleResize)
    }
  }, [expand])

  const statsToggleTransition = useTransition(!expand, null, {
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
  })

  return (
    <>
      {height && (
        <Spring
          to={{
            height,
          }}
        >
          {({ height: currentHeight }) =>
            renderHero({
              width,
              height: Math.trunc(currentHeight),
              targetHeight: height,
            })
          }
        </Spring>
      )}
      {statsToggleTransition.map(
        ({ item: collapsed, key, props }) =>
          collapsed && (
            <AnimatedBox key={key} style={props}>
              {renderContent()}
            </AnimatedBox>
          )
      )}
    </>
  )
}

export default Layout
