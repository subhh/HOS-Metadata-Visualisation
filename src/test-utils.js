import React, { useReducer } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { MemoryRouter } from 'react-router-dom'
import { StateContext, DispatchContext, reducer, initialState } from './state'
import theme from './theme'

const AllTheProviders = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <ThemeProvider theme={theme}>
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={dispatch}>
          {children}
        </DispatchContext.Provider>
      </StateContext.Provider>
    </ThemeProvider>
  )
}

const WithRouter = ({ children }) => (
  <MemoryRouter>
    <AllTheProviders>{children}</AllTheProviders>
  </MemoryRouter>
)

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export const customRenderWithRouter = (ui, options) =>
  render(ui, {
    wrapper: WithRouter,
    ...options,
  })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

export const mockAsync = data => {
  jest.spyOn(global, 'fetch').mockImplementation(url => {
    return Promise.resolve({
      json: () => Promise.resolve(data),
    })
  })
}

export const restoreAsyncMock = () => {
  global.fetch.mockRestore()
}
