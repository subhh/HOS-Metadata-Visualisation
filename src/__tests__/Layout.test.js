import React from 'react'
import Layout from 'components/Layout'
import { render, waitForElement } from 'test-utils'

const renderHero = ({ width, height, targetHeight }) => (
  <div>{`Hero ${width} ${height} ${targetHeight}`}</div>
)
const renderContent = () => <div data-testid="content">Content</div>

it('<Layout /> sould show Hero and Content when not expanded', async () => {
  const { asFragment, getByTestId } = render(
    <Layout renderHero={renderHero} renderContent={renderContent} />
  )

  await waitForElement(() => getByTestId('content'))
  expect(asFragment()).toMatchSnapshot()
})

it('<Layout /> should only show Hero when expanded', () => {
  const { container } = render(
    <Layout
      renderHero={renderHero}
      renderContent={renderContent}
      expand={true}
    />
  )
  expect(container.firstChild).toMatchSnapshot()
})
