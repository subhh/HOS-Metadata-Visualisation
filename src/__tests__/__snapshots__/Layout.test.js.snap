// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> should only show Hero when expanded 1`] = `
<div>
  Hero 1024 638 638
</div>
`;

exports[`<Layout /> sould show Hero and Content when not expanded 1`] = `
<DocumentFragment>
  <div>
    Hero 1024 307 307.20000000000005
  </div>
  <div
    class="sc-bdVaJa iHZvIS"
    style="opacity: 0;"
  >
    <div
      data-testid="content"
    >
      Content
    </div>
  </div>
</DocumentFragment>
`;
